### PR TITLE
Fix reflect ConstantPool bootstrapping issues

### PR DESCRIPTION
**Pull request by tajila**
We currently perform explicit native registration for jdk.internal.reflect.ConstantPool. This occurs after JCL init in VM startup. It is possible that ConstantPool will be used in JCL init before the natives are registered, this will result in signal 218 since we have public JNI stubs as a catch all for these capabilities.

I propose the following:

Step 1) (this PR)
- introduce a registerNative call in ConstantPool.

Step 2)
- update the JCL code to use the registerNative() in the static initializer of the class, thereby ensuring that the natives are registered before the class is used.

Step 3)
- remove the explicit register native call in JVM initialization

With these steps we dont need to do any build coordination between openj9 and the extension repos.

### CODE REVIEW

**Comment by tajila** [comment]
jenkins test sanity alinux64 jdk21

**Comment by JasonFengJ9** [comment]
> With these steps we dont need to do any build coordination between openj9 and the extension repos.

Assuming the proposed extension changes are in extension repo `openj9` branch, there is no acceptance build or manual promotion required like OMR contents.
I think it is efficient to make changes in both OpenJ9 and extension repos, and merge them at the same time (very small time windows to break some PR/personal builds) unless @keithc-ca or @pshipton disagree.

**Comment by pshipton** [comment]
I'm happy to go either way. We can merge it all at once, or stage it like Tobi suggests. Staging it makes it easier to test & merge version by version.

**Comment by keithc-ca** [comment]
> the proposed extension changes

Which changes are those?

File: runtime/jcl/common/sun_reflect_ConstantPool.c [hunk 1]
```diff
@@ -919,6 +927,12 @@ registerJdkInternalReflectConstantPoolNatives(JNIEnv *env) {
 	/* jdk.internal.reflect.ConstantPool is currently cached in CLS_sun_reflect_ConstantPool */
 	jclass jdk_internal_reflect_ConstantPool = JCL_CACHE_GET(env, CLS_sun_reflect_ConstantPool);
 
+	/* This is temporary code, will be removed in the future */
```
Comment (keithc-ca) [inline runtime/jcl/common/sun_reflect_ConstantPool.c:930]: Please finish the sentence with a period.

File: runtime/jcl/common/sun_reflect_ConstantPool.c [hunk 2]
```diff
@@ -919,6 +927,12 @@ registerJdkInternalReflectConstantPoolNatives(JNIEnv *env) {
...
+	if (J9_ARE_ANY_BITS_SET(vm->extendedRuntimeFlags2, J9_EXTENDED_RUNTIME2_REFLECT_CONSTANTPOOL_REGISTER_NATIVES_CALLED)) {
+		goto _end;
+	}
+	vm->extendedRuntimeFlags2 |= J9_EXTENDED_RUNTIME2_REFLECT_CONSTANTPOOL_REGISTER_NATIVES_CALLED;
```
Comment (keithc-ca) [inline runtime/jcl/common/sun_reflect_ConstantPool.c:934]: What ensures this is thread-safe?

**Review by keithc-ca — CHANGES_REQUESTED** [review]

File: runtime/jcl/common/sun_reflect_ConstantPool.c (hunk 2 above)
Comment (tajila) [reply runtime/jcl/common/sun_reflect_ConstantPool.c:934]: JVM init is single threaded, so is class init.

**Comment by tajila** [comment]
> Won't we end up in the same situation we have now?

No, this refers to the explicit call we are doing in JVM startup

**Comment by keithc-ca** [comment]
> No, this refers to the explicit call we are doing in JVM initialization

Are you referring to [this call in stdinit.c](https://github.com/eclipse-openj9/openj9/blob/master/runtime/jcl/common/stdinit.c#L235)?

**Comment by tajila** [comment]
> Are you referring to [this call in stdinit.c](https://github.com/eclipse-openj9/openj9/blob/master/runtime/jcl/common/stdinit.c#L235)?

yes

**Comment by tajila** [comment]
@keithc-ca any further comments?

**Review by keithc-ca — APPROVED** [review]

**Review by babsingh — APPROVED** [review]

**Review by JasonFengJ9 — APPROVED** [review]

**Comment by babsingh** [comment]
Old builds (all passed):
- JDK21 alinux64: https://openj9-jenkins.osuosl.org/view/Pull%20Requests/job/PullRequest-OpenJ9/4424
- JDK11 win: https://openj9-jenkins.osuosl.org/view/Pull%20Requests/job/PullRequest-OpenJ9/4423

IMO, the first set of builds should still be good. @tajila Do you recommend relaunching the builds?

**Comment by tajila** [comment]
Ill do a sanity compile since some source code was changed

**Comment by tajila** [comment]
jenkins compile alinux64 jdk11

**Pull request merged**